ouroboros-consensus/src/Ouroboros/Consensus/Protocol/TPraos.hs

```diff
@@ -0,0 +1,351 @@
+      ) => HeaderSupportsTPraos c hdr where
+
+  -- Because we are using the executable spec, rather than implementing the
+  -- protocol directly here, we have a fixed header type rather than an
+  -- abstraction. So we must introduce this method.
+  headerToBHeader
+    :: proxy (TPraos c)
+    -> hdr
+    -> BHeader c
+
+forgeTPraosFields :: ( HasNodeState (TPraos cfg c) m
+                    , MonadRandom m
+                    , TPraosCrypto c
+                    , Cardano.Crypto.KES.Class.Signable (KES c) toSign
+                    )
+                 => NodeConfig (TPraos cfg c)
+                 -> TPraosProof c
+                 -> (TPraosToSign c -> toSign)
+                 -> m (TPraosFields c toSign)
+forgeTPraosFields TPraosNodeConfig{..}  TPraosProof{..} mkToSign = do
+    let icn@TPraosIsCoreNode{..} = tpraosIsCoreNode
+        (DiscVKey issuerVK) = ocertVkCold tpraosIsCoreNodeOpCert
+        signedFields = TPraosToSign {
+          tptsIssuerVK = issuerVK
+        , tptsVrfVK = deriveVerKeyVRF tpraosIsCoreNodeSignKeyVRF
+        , tptsEta = tpraosEta
+        , tptsLeader = tpraosLeader
+        , tptsOCert = tpraosIsCoreNodeOpCert
+        }
+    m <- signedKES
+          ()
+          (fromIntegral (unSlotNo tpraosProofSlot))
+          (mkToSign signedFields)
+          tpraosIsCoreNodeSKSHot
+    case m of
+      Nothing                  -> error "mkOutoborosPayload: signedKES failed"
+      Just (signature, newKey) -> do
+        putNodeState . Just $ icn { tpraosIsCoreNodeSKSHot = newKey }
+        return $ TPraosFields {
+            tpraosSignature    = signature
+          , tpraosToSign       = (mkToSign signedFields)
+          }
+
+{-------------------------------------------------------------------------------
+  TPraos specific types
+-------------------------------------------------------------------------------}
+
+-- | Assembled proof that the issuer has the right to issue a block in the
+-- selected slot.
+data TPraosProof c
+  = TPraosProof
+    { tpraosEta        :: CertifiedVRF (VRF c) Nonce
+    , tpraosLeader     :: CertifiedVRF (VRF c) UnitInterval
+    , tpraosProofSlot  :: SlotNo
+    , tpraosIsCoreNode :: TPraosIsCoreNode c
+    } deriving Generic
+
+instance TPraosCrypto c => NoUnexpectedThunks (TPraosProof c)
+
+{-------------------------------------------------------------------------------
+  Protocol proper
+-------------------------------------------------------------------------------}
+
+data TPraos cfg c
+
+-- | TPraos parameters that are node independent
+data TPraosParams = TPraosParams {
+      -- | Active slots coefficient. This parameter represents the proportion of
+      -- slots in which blocks should be issued. This can be interpreted as the
+      -- probability that a party holding all the stake will be elected as
+      -- leader for a given slot.
+      tpraosLeaderF       :: Double
+    , tpraosSecurityParam :: SecurityParam
+    , tpraosEpochInfo     :: EpochInfo Identity
```

Comment:
I'm commenting on this here, but maybe "here" is just one occurrence of many, and we need to fix this in a more fundamental way. This `EpochInfo Identity` is problematic. It means that we must know, _statically_, the length of every epoch (in slots). _We don't know that_. Even if every ledger (Shelley, Byron, ..) has a static number of slots per epoch (i.e., it's not an updatable protocol parameters, the exact _time_ of the hard fork _does_ depend on the ledger and so we cannot produce an `EpochInfo Identity`. This is wrecking my head a little however. I guess we can merge this as is, and re-think `EpochInfo` when it comes to the hard fork combinator.